File: src/data/sensors/DeviceSensorProvider.ts
import type { TelemetryPoint } from '../../domain/models/telemetry';
import { pedometerEngine } from '../../services/pedometerEngine';

export class DeviceSensorProvider {
  private watchId: number | null = null;
  private isMotionActive = false;
  private callback: ((point: Partial<TelemetryPoint>) => void) | null = null;
  
  private lastAccelX = 0;
  private lastAccelY = 9.81;
  private lastAccelZ = 0;
  private lastGyroZ = 0;

  public async requestPermissions(): Promise<boolean> {
    try {
      // iOS DeviceMotionEvent permission check
      if (
        typeof DeviceMotionEvent !== 'undefined' &&
        typeof (DeviceMotionEvent as any).requestPermission === 'function'
      ) {
        const permission = await (DeviceMotionEvent as any).requestPermission();
        if (permission !== 'granted') {
          console.warn('DeviceMotion permission denied by user');
          return false;
        }
      }

      return true;
    } catch (e) {
      console.warn('Sensor permission error:', e);
      return false;
    }
  }

  public startListening(onPoint: (point: Partial<TelemetryPoint>) => void): boolean {
    this.callback = onPoint;

    // 1. Device Motion Listener (Hardware Accelerometer Peak Pedometer & Gyroscope)
    if (typeof window !== 'undefined' && 'DeviceMotionEvent' in window) {
      window.addEventListener('devicemotion', this.handleMotion, true);
      this.isMotionActive = true;
    }

    // 2. Geolocation Listener (GPS Speed & Coordinates)
    if (typeof navigator !== 'undefined' && 'geolocation' in navigator) {
      this.watchId = navigator.geolocation.watchPosition(
        this.handlePosition,
        (err) => console.warn('Geolocation error:', err.message),
        {
          enableHighAccuracy: true,
          maximumAge: 500,
          timeout: 5000,
        }
      );
      return true;
    }

    return false;
  }

  private handleMotion = (event: DeviceMotionEvent) => {
    const acc = event.accelerationIncludingGravity || event.acceleration;
    const rot = event.rotationRate;

    if (acc) {
      this.lastAccelX = parseFloat((acc.x || 0).toFixed(2));
      this.lastAccelY = parseFloat((acc.y || 9.81).toFixed(2));
      this.lastAccelZ = parseFloat((acc.z || 0).toFixed(2));

      // Feed motion vector into real-time pedometer engine
      pedometerEngine.processMotion(this.lastAccelX, this.lastAccelY, this.lastAccelZ);
    }

    if (rot) {
      this.lastGyroZ = parseFloat((rot.gamma || rot.alpha || 0).toFixed(1));
    }
  };

  private handlePosition = (pos: GeolocationPosition) => {
    if (!this.callback) return;

    const speedMps = pos.coords.speed || 0;
    const speedKmH = parseFloat((speedMps * 3.6).toFixed(1));

    this.callback({
      timestamp: pos.timestamp || Date.now(),
      latitude: pos.coords.latitude,
      longitude: pos.coords.longitude,
      altitude: pos.coords.altitude || 0,
      speedKmH,
      heading: pos.coords.heading || 0,
      accuracy: pos.coords.accuracy || 5,
      accelX: this.lastAccelX,
      accelY: this.lastAccelY,
      accelZ: this.lastAccelZ,
      gyroX: 0,
      gyroY: 0,
      gyroZ: this.lastGyroZ,
    });
  };

  public stopListening(): void {
    if (typeof window !== 'undefined' && this.isMotionActive) {
      window.removeEventListener('devicemotion', this.handleMotion, true);
      this.isMotionActive = false;
    }

    if (typeof navigator !== 'undefined' && this.watchId !== null) {
      navigator.geolocation.clearWatch(this.watchId);
      this.watchId = null;
    }

    this.callback = null;
  }
}

File: smart_trip_ai_mobile/lib/data/sensors/device_sensor_provider.dart
import 'dart:async';
import 'package:sensors_plus/sensors_plus.dart';
import 'package:geolocator/geolocator.dart';
import '../../domain/models/telemetry.dart';
import '../../services/pedometer_engine.dart';

class DeviceSensorProvider {
  StreamSubscription<Position>? _positionStream;
  StreamSubscription<UserAccelerometerEvent>? _accelStream;
  StreamSubscription<GyroscopeEvent>? _gyroStream;
  
  bool isMotionActive = false;
  Function(TelemetryPoint)? callback;
  
  double lastAccelX = 0;
  double lastAccelY = 9.81;
  double lastAccelZ = 0;
  double lastGyroZ = 0;

  Future<bool> requestPermissions() async {
    try {
      bool serviceEnabled = await Geolocator.isLocationServiceEnabled();
      if (!serviceEnabled) {
        return false;
      }

      LocationPermission permission = await Geolocator.checkPermission();
      if (permission == LocationPermission.denied) {
        permission = await Geolocator.requestPermission();
        if (permission == LocationPermission.denied) {
          return false;
        }
      }

      if (permission == LocationPermission.deniedForever) {
        return false;
      }
      return true;
    } catch (e) {
      print('Sensor permission error: $e');
      return false;
    }
  }

  Future<bool> startListening(Function(TelemetryPoint) onPoint) async {
    callback = onPoint;

    // 1. Device Motion Listener (Hardware Accelerometer Peak Pedometer & Gyroscope)
    try {
      _accelStream = userAccelerometerEventStream().listen((UserAccelerometerEvent event) {
        lastAccelX = double.parse(event.x.toStringAsFixed(2));
        lastAccelY = double.parse((event.y + 9.81).toStringAsFixed(2)); // adding approx gravity
        lastAccelZ = double.parse(event.z.toStringAsFixed(2));
        pedometerEngine.processMotion(lastAccelX, lastAccelY, lastAccelZ);
      });
      _gyroStream = gyroscopeEventStream().listen((GyroscopeEvent event) {
        lastGyroZ = double.parse(event.z.toStringAsFixed(1));
      });
      isMotionActive = true;
    } catch (e) {
      print('Sensors error: $e');
    }

    // 2. Geolocation Listener (GPS Speed & Coordinates)
    try {
      _positionStream = Geolocator.getPositionStream(
        locationSettings: const LocationSettings(
          accuracy: LocationAccuracy.high,
          distanceFilter: 0,
        ),
      ).listen(_handlePosition);
      return true;
    } catch (e) {
      print('Geolocator error: $e');
    }

    return false;
  }

  void _handlePosition(Position pos) {
    if (callback == null) return;

    final speedMps = pos.speed;
    final speedKmH = double.parse((speedMps * 3.6).toStringAsFixed(1));

    callback!(TelemetryPoint(
      id: 'pt-${DateTime.now().millisecondsSinceEpoch}',
      tripId: 'real-trip',
      timestamp: pos.timestamp?.millisecondsSinceEpoch ?? DateTime.now().millisecondsSinceEpoch,
      latitude: pos.latitude,
      longitude: pos.longitude,
      altitude: pos.altitude,
      speedKmH: speedKmH,
      heading: pos.heading,
      accuracy: pos.accuracy,
      accelX: lastAccelX,
      accelY: lastAccelY,
      accelZ: lastAccelZ,
      gyroX: 0,
      gyroY: 0,
      gyroZ: lastGyroZ,
      state: speedKmH > 3 ? DrivingState.driving : DrivingState.idling, // Simplified state derivation for raw sensors
      gForceCombined: double.parse((1.0 + (lastAccelZ.abs()) / 9.81).toStringAsFixed(2)),
    ));
  }

  void stopListening() {
    _accelStream?.cancel();
    _accelStream = null;
    
    _gyroStream?.cancel();
    _gyroStream = null;
    
    isMotionActive = false;

    _positionStream?.cancel();
    _positionStream = null;

    callback = null;
  }
}
